### PR TITLE
cranelift: Remove booleans

### LINKED ISSUE #3205: Cranelift: formalize, document, and audit binary representation of booleans
In #3202 and a number of past issues as well, we've repeatedly encountered some uncertainty as to (i) how boolean values of varying widths are stored in registers, (ii) whether they can be loaded/stored or not and what representation they take in memory (see e.g. #3102 where we addressed this for trampoline args), and (iii) how this interacts with boolean<->integer conversion ops like `bint` and `bmask`.

The understanding we seem to generally agree on is that booleans of width > 1 bit have all bits set (true) or all bits cleared (false). And a `b1` is stored in a register with the same invariant as other integer values: the upper bits (in this case everything above the LSB) are undefined.

However, this is far from being confusion-free and unambiguous; thus, it seems reasonable to (i) agree definitively on how bools are represented in registers and memory, and (ii) audit our compliance to whatever invariants we decide on.

### PR DESCRIPTION
**Pull request by elliottt**
Remove the boolean types from cranelift, and the associated instructions `breduce`, `bextend`, `bconst`, and `bint`. Standardize on using `1`/`0` for the return value from instructions that produce scalar boolean results, and `-1`/`0` for boolean vector elements.

Fixes #3205

### CODE REVIEW

File: cranelift/codegen/src/ir/extfunc.rs [hunk 1]
```diff
@@ -424,7 +424,5 @@ mod tests {
         assert_eq!(sig.to_string(), "(i32) -> f32 windows_fastcall");
         sig.params.push(AbiParam::new(I32.by(4).unwrap()));
         assert_eq!(sig.to_string(), "(i32, i32x4) -> f32 windows_fastcall");
-        sig.returns.push(AbiParam::new(B8));
```
Comment (bjorn3) [inline cranelift/codegen/src/ir/extfunc.rs:427]: Please push another type instead. I believe it tests multi return function signatures.

File: cranelift/codegen/src/isa/aarch64/lower/isle.rs [hunk 2]
```diff
@@ -164,7 +164,6 @@ impl Context for IsleContext<'_, '_, MInst, Flags, IsaFlags, 6> {
     fn integral_ty(&mut self, ty: Type) -> Option<Type> {
         match ty {
             I8 | I16 | I32 | I64 | R64 => Some(ty),
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/aarch64/lower/isle.rs:166]: This should list R32 too, right?

File: cranelift/filetests/filetests/runtests/bextend.clif [hunk 3]
```diff
@@ -5,85 +5,52 @@ target x86_64
...
+; run: %bextend_i8_i64(0) == 0
 
 
-function %bextend_b16_b32(b16) -> b32 {
-block0(v0: b16):
-  v1 = bextend.b32 v0
+function %bextend_i16_i32(i16) -> i32 {
+block0(v0: i16):
+  v1 = bextend.i32 v0
   return v1
```
Comment (bjorn3) [inline cranelift/filetests/filetests/runtests/bextend.clif:36]: Can bextend be replaced with uextend (for non-vector bools) and sextend (for vector bools)?

Comment (afonso360) [reply cranelift/filetests/filetests/runtests/bextend.clif:36]: I'd be in favor of removing `bextend` and `breduce`. The latter can also be replaced by `band_imm`.

**Comment by afonso360** [comment]
I'm really happy that we're finally doing this! Thank you for taking this on.

Some of our lowerings depend on the fact that bools can only be represented as all ones or all zeroes, and so now that we are removing that assumption we might want to revisit them.

The [AArch64 implementation for `bmask`](https://github.com/bytecodealliance/wasmtime/blob/e45577e097b064797def3554468cffa5fdd443d3/cranelift/codegen/src/isa/aarch64/lower.isle#L1556-L1560) is wrong if we want to accept any integer other than -1 and 0. (i.e. 2 will be false, but should be true)

`bint` is also assuming that inputs are all ones or all zeros and is [implemented as `and_imm` on AArch64](https://github.com/bytecodealliance/wasmtime/blob/e45577e097b064797def3554468cffa5fdd443d3/cranelift/codegen/src/isa/aarch64/lower.isle#L1541-L1542).

File: cranelift/codegen/src/ir/extfunc.rs (hunk 1 above)
Comment (elliottt) [reply cranelift/codegen/src/ir/extfunc.rs:427]: Thanks for catching that, I've switched it to `i8`.

File: cranelift/codegen/src/isa/aarch64/lower/isle.rs (hunk 2 above)
Comment (elliottt) [reply cranelift/codegen/src/isa/aarch64/lower/isle.rs:166]: Probably, though I'm not sure we should address that on this branch. Would you like to add it in a separate PR?

File: cranelift/codegen/meta/src/shared/instructions.rs [hunk 4]
```diff
@@ -685,7 +681,7 @@ pub(crate) fn define(
     let iflags: &TypeVar = &ValueType::Special(types::Flag::IFlags.into()).into();
     let fflags: &TypeVar = &ValueType::Special(types::Flag::FFlags.into()).into();
 
-    let b1: &TypeVar = &ValueType::from(LaneType::from(types::Bool::B1)).into();
+    let b1: &TypeVar = &ValueType::from(LaneType::from(types::Int::I8)).into();
```
Comment (cfallin) [inline cranelift/codegen/meta/src/shared/instructions.rs:684]: Do we still need a typevar named `b1` at all? Is this used for "truthy" things like icmp return values etc? If so, perhaps we could call it `truthy` or somesuch?

File: cranelift/codegen/src/ir/types.rs [hunk 5]
```diff
@@ -157,7 +152,7 @@ impl Type {
     /// Scalar types are all converted to `b1` which is usually what you want.
     pub fn as_bool(self) -> Self {
```
Comment (cfallin) [inline cranelift/codegen/src/ir/types.rs:153]: Perhaps we could remove the `.as_bool()` helper altogether?

File: cranelift/codegen/src/prelude.isle [hunk 6]
```diff
@@ -369,15 +357,15 @@
 (decl int_bool_fits_in_32 (Type) Type)
 (extern extractor int_bool_fits_in_32 int_bool_fits_in_32)
 
-;; An extractor that matches I64 or B64.
+;; An extractor that matches I64.
```
Comment (cfallin) [inline cranelift/codegen/src/prelude.isle:360]: We can probably remove this extractor now and just use a constant `$I64`, no? That has the nice extra benefit of allowing more precise overlap checking...

File: cranelift/codegen/src/verifier/mod.rs [hunk 7]
```diff
@@ -1536,28 +1535,6 @@ impl<'a> Verifier<'a> {
                             ));
                         }
                     }
-                    Opcode::Breduce | Opcode::Ireduce | Opcode::Fdemote => {
```
Comment (cfallin) [inline cranelift/codegen/src/verifier/mod.rs:1539]: I don't think we want to delete the `Ireduce` and `Fdemote` cases here?

File: cranelift/filetests/filetests/isa/aarch64/condbr.clif [hunk 8]
```diff
@@ -2,18 +2,18 @@ test compile precise-output
...
+function %f(i64, i64) -> i8 {
 block0(v0: i64, v1: i64):
   v2 = icmp eq v0, v1
   return v2
 }
 
 ; block0:
 ;   subs xzr, x0, x1
-;   cset x0, eq
+;   csetm x0, eq
```
Comment (cfallin) [inline cranelift/filetests/filetests/isa/aarch64/condbr.clif:13]: Is this right? I would expect an `icmp` to give us an `i8` with value either `0` or `1`, but here we're actually getting `0` or `-1` I think...

**Review by cfallin — COMMENTED** [review]
Wow -- thank you for doing the not-very-flashy but very important cleanup work here! It's surprising to me, but validating of this direction, that there are so many little corner cases that this removes.

Some scattered thoughts below; I read fairly carefully early in the diff, but my eyes started to glaze over partway through the test-output deltas, so I'll want to look at those some more in another pass. Hopefully this is good to get started though.

File: cranelift/filetests/filetests/runtests/bextend.clif (hunk 3 above)
Comment (cfallin) [reply cranelift/filetests/filetests/runtests/bextend.clif:36]: Yep, if we no longer have bools then these instructions no longer make sense; the user can zero- or sign-extend an integer (which happens to be a 0/1 or 0/-1 "truthy" value) in the usual ways.

**Comment by cfallin** [comment]
Also, just to note it here: if at all possible, I'd prefer for us to wait to merge this until after my mid-end work in #4953 is merged, if only because they're both large and likely-to-conflict PRs and mine has been through ~three painful rebases already :-)

File: cranelift/codegen/src/verifier/mod.rs (hunk 7 above)
Comment (elliottt) [reply cranelift/codegen/src/verifier/mod.rs:1539]: Good catch, I got a bit overzealous there :)

File: cranelift/codegen/src/ir/types.rs (hunk 5 above)
Comment (elliottt) [reply cranelift/codegen/src/ir/types.rs:153]: How would you feel about renaming it to `.as_truthy()` instead? It does convert types like `F32` to `I32`, so it would be convenient to keep something like it around.

File: cranelift/codegen/src/prelude.isle (hunk 6 above)
Comment (elliottt) [reply cranelift/codegen/src/prelude.isle:360]: That's a great point! I hadn't started thinking about improvements that we could make without the boolean types, and was just in "make it compile" mode instead :)

File: cranelift/codegen/meta/src/shared/instructions.rs (hunk 4 above)
Comment (elliottt) [reply cranelift/codegen/meta/src/shared/instructions.rs]: I renamed it to `i8`, as that should hopefully give good documentation for the places that it's used.

File: cranelift/filetests/filetests/isa/aarch64/condbr.clif (hunk 8 above)
Comment (elliottt) [reply cranelift/filetests/filetests/isa/aarch64/condbr.clif]: This was due to the implementation of `materialize_bool_result` in the aarch64 lowerings special-casing a result of `1` bit to produce values of `1`/`0`, and otherwise producing `-1`/`0`. I've removed the special case, and it now unconditionally produces `1`/`0`. The downside of that change is that it complicates the lowering of `bmask`, but I think that's a reasonable trade-off.

Comment (cfallin) [reply cranelift/filetests/filetests/isa/aarch64/condbr.clif]: Cool, makes sense, thanks!

Do we have any runtests that return the result of an `icmp` or friends and directly assert its value? Seems like if not, that would be good coverage: value is 0 or 1 for various cases for each kind of comparison (`icmp`, `fcmp`).

Comment (elliottt) [reply cranelift/filetests/filetests/isa/aarch64/condbr.clif]: We have a lot of tests that perform a comparison, `uextend` it to a larger type, and then test the resulting values. The `runtests/sqrt.clif` test is a good example of this.

Comment (cfallin) [reply cranelift/filetests/filetests/isa/aarch64/condbr.clif]: Awesome, we're covered then, thanks! (I guess I should just review the tests in general; still need to do that, sorry!)

File: cranelift/codegen/src/prelude.isle (hunk 6 above)
Comment (elliottt) [reply cranelift/codegen/src/prelude.isle]: I've made this change, and removed/refactored similar extractors in the prelude.

File: cranelift/filetests/filetests/isa/aarch64/condbr.clif (hunk 8 above)
Comment (elliottt) [reply cranelift/filetests/filetests/isa/aarch64/condbr.clif]: It might be worth refactoring those tests to return the value of the `icmp`/`fcmp` directly, as there might be interference from the `uextend`.

File: cranelift/filetests/filetests/runtests/bextend.clif (hunk 3 above)
Comment (elliottt) [reply cranelift/filetests/filetests/runtests/bextend.clif]: I really like using `uextend`/`sextend` in place of `bextend`, and have made that refactoring. I've also removed `bconst`, `bextend`, `breduce` and `bint`, which has collapsed a lot of cases in the test suite.

File: cranelift/codegen/src/isa/aarch64/inst.isle [hunk 9]
```diff
@@ -3449,78 +3442,115 @@
...
+;; registers in the 128-bit value, and then use csetm accordingly. This is
+;; slightly more efficient than two recursive calls to `lower_bmask`, as we can
+;; reuse the comparison of the hi register to select between the two results,
+;; and avoid an extra comparison.
+;;
+;; cmp   lo, #0
+;; csetm tmp1, ne
+;; cmp   hi, #0
+;; csetm tmp2, ne
+;; csel  dst, tmp1, tmp2, eq
```
Comment (afonso360) [inline cranelift/codegen/src/isa/aarch64/inst.isle:3481]: I was playing around with this, and I think I got something slightly shorter, what do you think about this:  https://github.com/bytecodealliance/wasmtime/commit/51baa19adfdabb9a31b0465c796ffad05bfaa0c3 ?

Comment (elliottt) [reply cranelift/codegen/src/isa/aarch64/inst.isle:3481]: I love it! Much better :)

Comment (elliottt) [reply cranelift/codegen/src/isa/aarch64/inst.isle:3481]: Would you like to push that commit to this branch, or should I make the change? Either way, I'll make sure to list you in the commit as `co-authored-by`.

Comment (elliottt) [reply cranelift/codegen/src/isa/aarch64/inst.isle]: Pushed, thanks Afonso!

**Comment by elliottt** [comment]
@uweigand, could you take a closer look at my changes to the s390x backend? In particular the changes in c01d860, as they increased the instruction count for the `bmask` translation by quite a bit.

File: cranelift/filetests/filetests/isa/aarch64/condops.clif [hunk 10]
```diff
@@ -763,14 +763,14 @@ block0(v0: i8, v1: i8, v2: i8):
...
-block0(v0: b1, v1: i8, v2: i8):
+function %i(i8, i8, i8) -> i8 {
+block0(v0: i8, v1: i8, v2: i8):
   v3 = select.i8 v0, v1, v2
   return v3
 }
 
 ; block0:
-;   and w5, w0, #1
+;   uxtb w5, w0
```
Comment (cfallin) [inline cranelift/filetests/filetests/isa/aarch64/condops.clif:773]: I think this could get a bit better (one instruction shorter) with

```
tst w0, #255
csel ...
```

(`tst w0, #255` is an alias for `ands wzr, w0, #255`; if we don't have machinst support for that yet it should be easy as another ALU op, one bit different from `and`)

File: cranelift/filetests/filetests/isa/x64/b1.clif [hunk 11]
```diff
@@ -1,8 +1,8 @@
 test compile precise-output
```
Comment (cfallin) [inline cranelift/filetests/filetests/isa/x64/b1.clif:1]: we can probably call this file something other than `b1.clif` now -- maybe `conditional_values.clif` or something?

File: cranelift/filetests/src/function_runner.rs [hunk 12]
```diff
@@ -425,42 +424,19 @@ fn make_trampoline(name: UserFuncName, signature: &ir::Signature, isa: &dyn Targ
...
-            let loaded = builder.ins().load(
-                ty,
+            builder.ins().load(
+                param.value_type,
                 flags,
                 values_vec_ptr_val,
                 (i * UnboxedValues::SLOT_SIZE) as i32,
-            );
-
-            // For booleans, we want to type-convert the loaded integer into a boolean and ensure
```
Comment (cfallin) [inline cranelift/filetests/src/function_runner.rs:445]: So nice to see this mess of canonicalization and type-coercion go away!

File: cranelift/reader/src/parser.rs [hunk 13]
```diff
@@ -2565,14 +2538,14 @@ impl<'a> Parser<'a> {
...
                 } else if sig.params.is_empty()
                     && sig.returns.len() == 1
-                    && sig.returns[0].value_type.is_bool()
+                    && sig.returns[0].value_type.is_int()
                 {
                     // To match the existing run behavior that does not require an explicit
-                    // invocation, we create an invocation from a function like `() -> b*` and
-                    // compare it to `true`.
+                    // invocation, we create an invocation from a function like `() -> i*` and
+                    // compare it to not `false`.
```
Comment (cfallin) [inline cranelift/reader/src/parser.rs:2545]: s/false/0/ here (or "and require the return value to be nonzero")?

**Review by cfallin — APPROVED** [review]
A few miscellaneous comments below, but this looks great to me after looking over the test deltas too. Thanks so much for the detailed and careful work here!

File: cranelift/reader/src/parser.rs (hunk 13 above)
Comment (elliottt) [reply cranelift/reader/src/parser.rs:2545]: Great point. I like the latter phrasing, and will make that change.

File: cranelift/filetests/filetests/isa/x64/b1.clif (hunk 11 above)
Comment (elliottt) [reply cranelift/filetests/filetests/isa/x64/b1.clif:1]: Good point, the existing name isn't really all that meaningful going forward :)

File: cranelift/filetests/filetests/isa/aarch64/condops.clif (hunk 10 above)
Comment (elliottt) [reply cranelift/filetests/filetests/isa/aarch64/condops.clif:773]: Would you be okay with addressing this one in follow-up PR?

Comment (cfallin) [reply cranelift/filetests/filetests/isa/aarch64/condops.clif:773]: Yep, that's fine!

Comment (elliottt) [reply cranelift/filetests/filetests/isa/aarch64/condops.clif:773]: I went ahead and fixed it on this branch, and found a bug with the x64 lowering of `select` in the process!

File: cranelift/filetests/filetests/runtests/br.clif [hunk 14]
```diff
@@ -156,3 +157,4 @@ block2:
 ; run: %brnz_i8(0) == 0
 ; run: %brnz_i8(1) == 1
 ; run: %brnz_i8(-1) == 1
+; run: %brnz_i8(97) == 1
```
Comment (cfallin) [inline cranelift/filetests/filetests/runtests/br.clif:160]: while here, a useful runtest to add might be a very slight variation on this one that intentionally tries to have nonzero bits in the upper undef bits of the `i8`-in-register, to make sure we're really testing just the low 8 bits; for example, by adding and wrapping:

```
block0(v0: i8):
  v1 = iconst.i8 255
  v2 = iadd.i8 v0, v1
  brz v2, ...
```

then it should branch for input arg `v0 = 1` even though the register will likely contain `256`...

File: cranelift/filetests/filetests/isa/x64/conditional-values.clif [hunk 15]
```diff
@@ -10,7 +10,7 @@ block0(v0: i8, v1: i32, v2: i32):
 ;   pushq   %rbp
 ;   movq    %rsp, %rbp
 ; block0:
-;   testb   $1, %dil
+;   testb   %dil, %dil
```
Comment (cfallin) [inline cranelift/filetests/filetests/isa/x64/conditional-values.clif:13]: Good catch! Just to make sure, do we have a runtest somewhere that tests with a nonzero-but-even value to catch LSB-only checks like this?

**Review by cfallin — APPROVED** [review]
A few thoughts on latest changes (but still LGTM!)

File: cranelift/codegen/src/isa/s390x/inst.isle [hunk 16]
```diff
@@ -3478,13 +3475,14 @@
...
+;; "true" is -1 for all result types except for $I8, which uses 1.
 (decl lower_bool (Type ProducesBool) Reg)
-(rule (lower_bool $B1 cond) (select_bool_imm $B1 cond 1 0))
-(rule (lower_bool $B8 cond) (select_bool_imm $B8 cond -1 0))
-(rule (lower_bool $B16 cond) (select_bool_imm $B16 cond -1 0))
-(rule (lower_bool $B32 cond) (select_bool_imm $B32 cond -1 0))
-(rule (lower_bool $B64 cond) (select_bool_imm $B64 cond -1 0))
+(rule (lower_bool $I8 cond) (select_bool_imm $I8 cond 1 0))
+
+;; TODO: do we need these cases anymore if B8..B128 are missing?
```
Comment (uweigand) [inline cranelift/codegen/src/isa/s390x/inst.isle:3482]: Hmm.  This helper is only called for possible scalar result types of `icmp` and `fcmp`.   These are specified as `&Int.as_bool()` and `&Float.as_bool()` so I thought I needed to handle multiple widths here.  But looking at this again, actually the only scalar return type of `as_bool` has always been `$B1` (and is now `$I8`), so it seems this is only type we need to handle here in the first place.

**Comment by uweigand** [comment]
> @uweigand, could you take a closer look at my changes to the s390x backend? In particular the changes to the lowering of `bmask`, as they increased the instruction count for the translation by quite a bit.

Well, in part this is due to changed semantics: the old `bmask` would take a boolean type as input, which (except for `$B1`) already used only the values 0 and -1.  So the semantics of `bmask` was simply a no-op or a sign-extend.  The new `bmask` I understand now needs to take any arbitrary integer input, so it actually has to perform a comparison against zero.

That said, I think the new implementation could still be improved, in particular the `$I128` cases.  However, instead of re-implementing the wheel here, why don't you simply use the existing `icmp_val` helper to emit the actual comparison?

Maybe something alone the lines of
```
 (rule (lower (has_type oty (bmask x @ (value_type ity))))
       (lower_bool_to_mask oty (icmp_val $true (IntCC.NotEqual) x (imm ity 0)))
```
with a new helper `lower_bool_to_mask` analogous to the old `lower_bool` ...

**Comment by uweigand** [comment]
Following up on myself, it would be even simpler *and* better to use the existing `value_nonzero` helper, which will fold in previous comparisons if possible:
```
 (rule (lower (has_type ty (bmask x)))
       (lower_bool_to_mask ty (value_nonzero x)))
```

**Comment by elliottt** [comment]
> Following up on myself, it would be even simpler _and_ better to use the existing `value_nonzero` helper, which will fold in previous comparisons if possible:
> 
> ```
>  (rule (lower (has_type ty (bmask x)))
>        (lower_bool_to_mask ty (value_nonzero x)))
> ```

This is much nicer! I'll rework the changes to add `lower_bool_to_mask` instead.

File: cranelift/filetests/filetests/isa/x64/conditional-values.clif (hunk 15 above)
Comment (elliottt) [reply cranelift/filetests/filetests/isa/x64/conditional-values.clif:13]: We do in the runtests: there's a case that checks the return value with `2` as the argument to `select` in `runtests/br.clif`.

File: cranelift/filetests/filetests/runtests/br.clif (hunk 14 above)
Comment (elliottt) [reply cranelift/filetests/filetests/runtests/br.clif:160]: That's a great idea, I'll add that test!

**Comment by uweigand** [comment]
> This is much nicer! I'll rework the changes to add `lower_bool_to_mask` instead.

Thanks!   The s390x parts LGTM now.
